deepspeed/module_inject/replace_module.py

```diff
@@ -764,14 +767,40 @@ def replace_fn(child, _policy, layer_id=0):
                                      replace_fn=replace_fn,
                                      _replace_policy=policy)
 
+    start_time = time.time()
+    rank = dist.get_rank() if dist.is_initialized() else 0
     if checkpoint is not None:
-        pbar = tqdm.tqdm(total=len(checkpoint),
-                         desc=f"Loading {len(checkpoint)} checkpoint shards")
-        for i in range(len(checkpoint)):
-            if not deepspeed.comm.is_initialized() or deepspeed.comm.get_rank() == 0:
-                pbar.update(1)
-            sd = torch.load(checkpoint[i], map_location='cpu')
-            load_model_with_checkpoint(replaced_module, sd, mp_replace)
+
+        if ckpt_type == 'pp':
+            pbar = tqdm.tqdm(total=len(checkpoint),
+                             desc=f"Loading {len(checkpoint)} checkpoint shards")
+            for i in range(len(checkpoint)):
+                if not deepspeed.comm.is_initialized() or deepspeed.comm.get_rank() == 0:
+                    pbar.update(1)
+                sd = torch.load(checkpoint[i], map_location='cpu')
+                load_model_with_checkpoint(replaced_module, sd, mp_replace, ckpt_type)
+        else:
+            world_size = dist.get_world_size() if dist.is_initialized() else 1
+            num_checkpoints = len(checkpoint) // world_size
+            assert world_size >= len(checkpoint),\
+                "Currently, merging checkpoints is not supported (when world_size is smaller than #checkpoints)!"
+            checkpoint_stride = world_size // len(checkpoint)
+            pbar = tqdm.tqdm(total=num_checkpoints,
+                             desc=f"Loading {num_checkpoints} checkpoint shards")
+            for i in range(num_checkpoints):
+                if not deepspeed.comm.is_initialized() or deepspeed.comm.get_rank() == 0:
+                    pbar.update(1)
+                sd = torch.load(checkpoint[i * world_size + (rank // checkpoint_stride)],
+                                map_location='cpu')
+                load_model_with_checkpoint(replaced_module, sd, mp_replace, ckpt_type)
+    print(f"checkpoint loading time at rank {rank}: {time.time()-start_time} sec")
+
+    if save_mp_checkpoint_path is not None:
+        if dist.is_initialized():
+            dist.barrier()
+        torch.save(replaced_module.transformer.state_dict(),
```

Comment:
How does the process of loading from this checkpoint after it is saved work?
Does it require a code change to point to the newly saved path?

Reply:
That's a good question. We currently need the user to pass the config json file that includes the list of checkpoint files, when loading the saved checkpoint. This can be passed through `checkpoint` argument using `deepspeed.init_inference`. Here is an example:
```
model = deepspeed.init_inference(model, 
                                 mp_size=world_size,
                                 dtype=torch.half,
                                 checkpoint='BLOOM-350m.json',
                                 replace_with_kernel_inject=True,
                                 )
```
I also add new part here at saving checkpoint that creates this json file, so you don't need to create that.

Reply:
I see the change you made, so the first time we would pass in the checkpoints.json with the list of huggingface checkpoints and a dir for `save_mp_checkpoint_path`. Then after that runs we can just pass the path to the newly created json file inside  `save_mp_checkpoint_path` as the `checkpoint` arg in `init_inference()`?

Reply:
Yes, that is correct :)